patches/minecraft/net/minecraft/data/TagsProvider.java.patch

```diff
@@ -1,34 +1,56 @@
 --- a/net/minecraft/data/TagsProvider.java
 +++ b/net/minecraft/data/TagsProvider.java
-@@ -28,11 +28,17 @@
+@@ -28,11 +28,19 @@
     protected final DataGenerator field_200433_a;
     protected final Registry<T> field_200435_c;
     protected final Map<ResourceLocation, ITag.Builder> field_200434_b = Maps.newLinkedHashMap();
 +   protected String modId;
++   net.minecraftforge.common.data.ExistingFileHelper existingFileHelper;
  
 +   @Deprecated//Forge, Use ModID version.
     protected TagsProvider(DataGenerator p_i49827_1_, Registry<T> p_i49827_2_) {
 -      this.field_200433_a = p_i49827_1_;
 -      this.field_200435_c = p_i49827_2_;
-+       this(p_i49827_1_, p_i49827_2_, "vanilla");
++       this(p_i49827_1_, p_i49827_2_, "vanilla", null);
     }
-+   protected TagsProvider(DataGenerator generatorIn, Registry<T> registryIn, String modId) {
++   protected TagsProvider(DataGenerator generatorIn, Registry<T> registryIn, String modId, @javax.annotation.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
 +      this.field_200433_a = generatorIn;
 +      this.field_200435_c = registryIn;
 +      this.modId = modId;
++      this.existingFileHelper = existingFileHelper;
 +   }
  
     protected abstract void func_200432_c();
  
-@@ -53,6 +59,7 @@
-          } else {
+@@ -49,10 +57,17 @@
+       this.field_200434_b.forEach((p_240524_4_, p_240524_5_) -> {
+          List<ITag.Proxy> list = p_240524_5_.func_232963_b_(function, function1).collect(Collectors.toList());
+          if (!list.isEmpty()) {
++            // FORGE: Add validation via existing resources
++            for (ITag.Proxy missingTag : list) {
++               // Vanilla tag types use plural names, revisit this when/if we add custom tag types
++               if (existingFileHelper == null || !missingTag.func_232968_a_().exists(((Registry)Registry.field_212617_f).func_177774_c(field_200435_c).func_110623_a() + "s", existingFileHelper))
+             throw new IllegalArgumentException(String.format("Couldn't define tag %s as it is missing following references: %s", p_240524_4_, list.stream().map(Objects::toString).collect(Collectors.joining(","))));
+-         } else {
++            }
++         } //else {
++            // FORGE: Run this code always, it will be skipped if the above code throws an exception
              JsonObject jsonobject = p_240524_5_.func_232965_c_();
              Path path = this.func_200431_a(p_240524_4_);
 +            if (path == null) return; //Forge: Allow running this data provider without writing it. Recipe provider needs valid tags.
  
              try {
                 String s = field_200437_e.toJson((JsonElement)jsonobject);
-@@ -78,7 +85,7 @@
+@@ -70,7 +85,7 @@
+                field_200436_d.error("Couldn't save tags to {}", path, ioexception);
+             }
+ 
+-         }
++//         }
```

Comment:
I would just remove the else and keep the brackets, as this commented out curly bracket is floating around quite freely otherwise (which might cause issues when porting patches)